core/src/main/java/org/apache/accumulo/core/fate/Fate.java

```diff
@@ -413,111 +324,81 @@ public void run() {
-              log.warn(
-                  "All Fate threads appear to be busy for the last {} minutes,"
-                      + " consider increasing property: {}",
-                  interval, Property.MANAGER_FATE_THREADPOOL_SIZE.getKey());
-              // Clear the history so that we don't log for interval minutes.
-              idleCountHistory.clear();
-            } else {
-              while (idleCountHistory.size() >= interval * 2) {
-                idleCountHistory.remove();
-              }
-            }
-          }
-          idleCountHistory.add(workQueue.getWaitingConsumerCount());
-        }
-      }
-    }, INITIAL_DELAY.toSeconds(), getPoolWatcherDelay().toSeconds(), SECONDS));
-    this.transactionExecutor = pool;
+    ThreadPools.watchCriticalScheduledTask(
+        fatePoolsWatcher.scheduleWithFixedDelay(new FatePoolsWatcher(environment, conf),
+            INITIAL_DELAY.toSeconds(), getPoolWatcherDelay().toSeconds(), SECONDS));
 
     ScheduledExecutorService deadResCleanerExecutor = null;
     if (runDeadResCleaner) {
       // Create a dead reservation cleaner for this store that will periodically clean up
       // reservations held by dead processes, if they exist.
       deadResCleanerExecutor = ThreadPools.getServerThreadPools().createScheduledExecutorService(1,
-          store.type() + "-dead-reservation-cleaner-pool");
+          store.type() == FateInstanceType.USER ? USER_DEAD_RESERVATION_CLEANER_POOL.poolName
+              : META_DEAD_RESERVATION_CLEANER_POOL.poolName);
       ScheduledFuture<?> deadReservationCleaner =
           deadResCleanerExecutor.scheduleWithFixedDelay(new DeadReservationCleaner(),
               INITIAL_DELAY.toSeconds(), getDeadResCleanupDelay().toSeconds(), SECONDS);
       ThreadPools.watchCriticalScheduledTask(deadReservationCleaner);
     }
     this.deadResCleanerExecutor = deadResCleanerExecutor;
 
-    this.workFinder = Threads.createThread("Fate work finder", new WorkFinder());
-    this.workFinder.start();
+    startFateExecutors(environment, conf, fateExecutors);
+  }
+
+  protected void startFateExecutors(T environment, AccumuloConfiguration conf,
+      Set<FateExecutor<T>> fateExecutors) {
+    for (var poolConf : getPoolConfigurations(conf).entrySet()) {
+      // no fate threads are running at this point; fine not to synchronize
+      fateExecutors
+          .add(new FateExecutor<>(this, environment, poolConf.getKey(), poolConf.getValue()));
+    }
+  }
+
+  /**
+   * Returns a map of the current pool configurations as set in the given config. Each key is a set
+   * of fate operations and each value is an integer for the number of threads assigned to work
+   * those fate operations.
+   */
+  protected Map<Set<FateOperation>,Integer> getPoolConfigurations(AccumuloConfiguration conf) {
+    Map<Set<FateOperation>,Integer> poolConfigs = new HashMap<>();
+    final var json = JsonParser.parseString(conf.get(getFateConfigProp())).getAsJsonObject();
+
```

Comment:
The code in PropertyType validates that all FateOperatiopns are present in the config that there is no duplication.  Not sure where this should go, but in the case where those conditions are not met will it get to this code?

Also it would be nice to log somewhere what the problem, if that is not being done.  For example log that  COMMIT_COMPACTION was not seen in the config, or log that SYSTEM_SPLIT was seen twice in the config.  Not sure where that should be done, was thinking that maybe it could be done here.

Reply:
If PropertyType acts how I expect, it should not reach this code. My interpretation was that PropertyType is how the property is tested for validity before ever being set ensuring an invalid prop can never be set. If this is not how property type works, it may reach this code. (I'm not sure what other purpose PropertyType would serve if not this, though).

Added logs to PropertyType to explain the reason for failure-to-set.